driver-core/src/main/com/mongodb/client/model/expressions/SwitchExpression.java

```diff
@@ -0,0 +1,117 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import com.mongodb.annotations.Immutable;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.util.Collections.unmodifiableList;
+
+/**
+ * Evaluates a series of case expressions. When it finds an expression which evaluates to true, $switch executes a specified expression
+ * and breaks out of the control flow.
+ *
+ * @since 4.?
+ */
+@Immutable
+public final class SwitchExpression implements Expression {
+    private final List<Branch> branches;
+    @Nullable
+    private final Expression defaultExpr;
+
+    SwitchExpression(final List<Branch> branches, @Nullable final Expression defaultExpr) {
+        this.branches = notNull("branches", branches);
+        this.defaultExpr = defaultExpr;
+    }
+
+    /**
+     * The path to take if no branch case expression evaluates to true.
+     *
+     * @param defaultExpr an expression to evaluate if no branch evaluates to true
+     * @return a new SwitchExpression
+     */
+    public SwitchExpression defaultExpr(final Expression defaultExpr) {
+        return new SwitchExpression(branches, defaultExpr);
+    }
+
+    /**
+     * An array of control branch documents.
+     *
+     * @return a list of branches, which may not be null.
+     */
+    public List<Branch> getBranches() {
+        return unmodifiableList(branches);
+    }
+
+    /**
+     * The path to take if no branch evaluates to true.
+     *
+     * @return an expression to evaluate if no branch evaluates to true. This may be null
+     */
+    @Nullable
+    public Expression getDefault() {
+        return defaultExpr;
+    }
+
+    @Override
+    public BsonValue toBsonValue(final CodecRegistry codecRegistry) {
+        BsonDocument switchDocument = new BsonDocument("branches",
+                branches.stream().map(branch ->
+                        new BsonDocument("case", branch.getCase().toBsonValue(codecRegistry))
+                                .append("then", branch.getThen().toBsonValue(codecRegistry)))
+                        .collect(Collectors.toCollection(BsonArray::new)));
+        if (defaultExpr != null) {
+            switchDocument.append("default", defaultExpr.toBsonValue(codecRegistry));
+        }
+        return new BsonDocument("$switch", switchDocument);
+    }
+
+    @Override
+    public String toString() {
+        return "SwitchExpression{" +
```

Comment:
Side note: how does everyone feel about turning off the checkstyle rules that make the generated toString and equals methods fail the check?  

 * `+` and `&&` on new line
 * braces required on `if` (this one I feel kind of tied to)

It's really annoying to have to modify all the generated method (I didn't do it yet, so checkstyle actually fails now)

Reply:
I'm fine with turning of the `+` and `&&` on a new line - I agree braces required on `if` could probably stay.

Reply:
Personally, I would be happy to just use the intellij defaults for these auto generated overrides.

Reply:
Ok, I'll see if there's a way to disable the checks just for these methods.